### ISSUE TITLE
Fails to parse existential type in statement position

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#52631]
As of rustc 1.29.0-nightly (874dec25e 2018-07-21) the following works:

```rust
#![feature(existential_type)]

use std::fmt::Debug;

existential type Existential: Debug;

fn main() {
    fn f() -> Existential {}
    println!("{:?}", f());
}
```

But the following does not parse. I would expect this to be equivalent to the above except for `Existential` in scope just within `main`, but if this is disallowed by the RFC then it should produce a better error message.

```rust
#![feature(existential_type)]

use std::fmt::Debug;

fn main() {
    existential type Existential: Debug;

    fn f() -> Existential {}
    println!("{:?}", f());
}
```

```console
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `type`
 --> src/main.rs:6:17
  |
6 |     existential type Existential: Debug;
  |                 ^^^^ expected one of 8 possible tokens here

error[E0425]: cannot find value `existential` in this scope
 --> src/main.rs:6:5
  |
6 |     existential type Existential: Debug;
  |     ^^^^^^^^^^^ not found in this scope
```

Mentioning the existential types tracking issue https://github.com/rust-lang/rust/issues/34511
Mentioning @oli-obk and @cramertj

**Comment by dtolnay** [comment]
Workaround: macros to the rescue.

```rust
#![feature(existential_type)]

use std::fmt::Debug;

macro_rules! this_is_an_item {
    ($i:item) => { $i };
}

fn main() {
    this_is_an_item! {
        existential type Existential: Debug;
    }

    fn f() -> Existential {}
    println!("{:?}", f());
}
```

**Comment by dtolnay** [comment]
I would like this to be fixed even if we know the syntax will change. Some use cases require existential type in statement position -- https://github.com/rust-lang/rust/issues/52632#issuecomment-406899110 is one example -- and we want people with those use cases to be able to try out existential type and provide feedback that informs the ultimate implementation.
